cobbler/module_loader.py

```diff
@@ -21,160 +20,168 @@
-    return value
-
-
-def get_module_from_file(
-    category: str, field: str, fallback_module_name: Optional[str] = None
-):
-    """
-    Get Python module, based on name defined in configuration file
-
-    :param category: field category in configuration file
-    :param field: field in configuration file
-    :param fallback_module_name: default value used if category/field is not found in configuration file
-    :raises CX: If unable to load Python module
-    :returns: A Python module.
-    """
-
-    module_name = get_module_name(category, field, fallback_module_name)
-    requested_module = MODULE_CACHE.get(module_name, None)
-    if requested_module is None:
-        raise CX("Failed to load module for %s/%s" % (category, field))
-    return requested_module
-
-
-def get_modules_in_category(category: str) -> list:
-    """
-    Return all modules of a module category.
-
-    :param category: The module category.
-    :return: A list of all modules of that category. Returns an empty list if the Category does not exist.
-    """
-    if category not in MODULES_BY_CATEGORY:
+        if module_path:
+            self.mod_path = module_path
+        self.module_cache: Dict[str, Any] = {}
+        self.modules_by_category: Dict[str, Dict[str, Any]] = {}
+        self.api = api
+
+    def load_modules(self):
+        """
+        Load the modules from the path handed to the function into Cobbler.
+
+        :return: Two dictionary's with the dynamically loaded modules.
+        """
+
+        filenames = glob.glob("%s/*.py" % self.mod_path)
+        filenames += glob.glob("%s/*.pyc" % self.mod_path)
+        filenames += glob.glob("%s/*.pyo" % self.mod_path)
+        # Allow recursive modules
+        filenames += glob.glob("%s/**/*.py" % self.mod_path)
+        filenames += glob.glob("%s/**/*.pyc" % self.mod_path)
+        filenames += glob.glob("%s/**/*.pyo" % self.mod_path)
+
+        for fn in filenames:
+            basename = fn.replace(self.mod_path, "")
+            modname = ""
+
+            if basename.__contains__("__pycache__") or basename.__contains__(
+                "__init__.py"
+            ):
+                continue
+
+            if basename[0] == "/":
+                basename = basename[1:]
+
+            basename = basename.replace("/", ".")
+
+            if basename[-3:] == ".py":
+                modname = basename[:-3]
+            elif basename[-4:] in [".pyc", ".pyo"]:
+                modname = basename[:-4]
+
+            self.__import_module(modname)
+
+        return self.module_cache, self.modules_by_category
+
+    def __import_module(self, modname: str):
+        """
+        Import a module which is not part of the core functionality of Cobbler.
+
+        :param modname: The name of the module.
+        """
+        try:
+            blip = import_module("cobbler.modules.%s" % modname)
+            if not hasattr(blip, "register"):
+                if not modname.startswith("__init__"):
```

Comment:
What's the deal with modules that are named "__init__abc"?

Reply:
So this was actually introduced by a commit author with the name "root" in commit 321084391b0115dd2b68b03f367a8f859717d3c6. I actually think it is redundant as I have added in 1e370ff954f1e966a36d66403b6387e1981f9103 the check at an earlier place. Soooo tldr: historically grown. I am not aware of modules with the name `__init__*`, however can I say that they don't exist? No sadly I can't...

Ideas if we should remove it or not?

Reply:
I'd remove it